### PR TITLE
Lower to index lambda refactor

### PR DESCRIPTION
**Pull request by kaushikcfd**
Trie to implement the suggestions in https://github.com/inducer/pytato/pull/375#issuecomment-1304895412

### CODE REVIEW

**Comment by kaushikcfd** [comment]
@inducer: I'm unable to make the MRO to pick the correct mapper methods. Other than that everything works here. Suggestions?

File: pytato/codegen.py [hunk 1]
```diff
@@ -85,7 +80,7 @@ def _generate_name_for_temp(
 
 # {{{ preprocessing for codegen
 
-class CodeGenPreprocessor(CopyMapper):
+class CodeGenPreprocessor(CopyMapper, ToIndexLambdaMapper):
```
Comment (inducer) [inline pytato/codegen.py:83]: Shouldn't `ToIndexLambdaMapper` come first?

Comment (kaushikcfd) [reply pytato/codegen.py:83]: That would make `CodeGenPreprocessor.rec == ToIndexLambdaMapper.rec`.

Comment (kaushikcfd) [reply pytato/codegen.py:83]: Aah I guess another abstract class. :facepalm: My bad.

Comment (kaushikcfd) [reply pytato/codegen.py]: This has been fixed!

**Comment by kaushikcfd** [comment]
This is a bit annoying to review as the <kbd>Tab</kbd> makes the `git-diff` very verbose. No updates were made to the mapper methods just moved pieces.

**Comment by inducer** [comment]
LGTM, thanks!

**Pull request merged**